pallets/committee/src/lib.rs

```diff
@@ -502,9 +541,32 @@ impl<T: Trait<I>, I: Instance> Module<T, I> {
             Self::deposit_event(event);
         }
 
-        // remove vote
-        <Voting<T, I>>::remove(&proposal);
-        <Proposals<T, I>>::mutate(|proposals| proposals.retain(|h| h != &proposal));
+        // Clear remaining proposal data.
+        Self::clear_proposal(&proposal);
+    }
+
+    /// Clear data for `proposal`, except for `ProposalOf`,
+    /// which needs to be cleared separately.
+    fn clear_proposal(proposal: &T::Hash) {
+        <Voting<T, I>>::remove(proposal);
+        <Proposals<T, I>>::mutate(|proposals| proposals.retain(|h| h != proposal));
+    }
+
+    /// Ensure that the given `proposal` with associated `expiry` hasn't expired relative to `now`.
+    /// As a side-effect, on error, any existing proposal data is pruned.
+    fn ensure_not_expired(
+        proposal: &T::Hash,
+        expiry: MaybeBlock<T::BlockNumber>,
+        now: T::BlockNumber,
+    ) -> Result<(), Error<T, I>> {
+        match expiry {
+            MaybeBlock::Some(e) if e <= now => {
+                Self::clear_proposal(proposal);
+                <ProposalOf<T, I>>::remove(proposal);
```

Comment:
Should we be updating storage and then returning an Err @maxsam4 ?

Reply:
So this is intended to be a lazy variant of the same "garbage collection" done when PIPs expire conditionally resulting in pruning (if historical pruning is enabled), as also implemented in this PR. Otherwise we could just leave the proposal be, but then it will stay there forever, which seems sub-optimal?

Erroring felt right to me, at least the first time (since pruning happens you can vote again after), since this gives the submitter a clearer indication that there's something wrong (instead of just silently resetting the vote tally).

Reply:
I think this is fine. Deleting expired proposals seems ok to me.